### PR TITLE
`Split*::as_str` refactor

### LINKED ISSUE #84974: `str::split_terminator` is never finished
I tried this code:

```rust
let mut iter = "A..B..".split_terminator('.');
dbg!(&iter); // finished: false (expected)
    
iter.by_ref().for_each(drop);
dbg!(&iter); // finished: false (unexpected)
    
dbg!(iter.next());
dbg!(&iter); // finished: false (unexpected)
```
<details><summary>output</summary>
<p>

```
[src/main.rs:3] &iter = SplitTerminator(
    SplitInternal {
        start: 0,
        end: 6,
        matcher: CharSearcher {
            haystack: "A..B..",
            finger: 0,
            finger_back: 6,
            needle: '.',
            utf8_size: 1,
            utf8_encoded: [
                46,
                0,
                0,
                0,
            ],
        },
        allow_trailing_empty: false,
        finished: false,
    },
)
[src/main.rs:6] &iter = SplitTerminator(
    SplitInternal {
        start: 6,
        end: 6,
        matcher: CharSearcher {
            haystack: "A..B..",
            finger: 6,
            finger_back: 6,
            needle: '.',
            utf8_size: 1,
            utf8_encoded: [
                46,
                0,
                0,
                0,
            ],
        },
        allow_trailing_empty: false,
        finished: false,
    },
)
[src/main.rs:8] iter.next() = None
[src/main.rs:9] &iter = SplitTerminator(
    SplitInternal {
        start: 6,
        end: 6,
        matcher: CharSearcher {
            haystack: "A..B..",
            finger: 6,
            finger_back: 6,
            needle: '.',
            utf8_size: 1,
            utf8_encoded: [
                46,
                0,
                0,
                0,
            ],
        },
        allow_trailing_empty: false,
        finished: false,
    },
)
```

</p>
</details>

I've expected the `finished` flag to be set when the iterator is exhausted, however it isn't.

### Meta

Tested on playground's [stable](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=46a4708ce0c82c8d86e1d3d610e7730f) `1.51.0` and [nightly](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=46a4708ce0c82c8d86e1d3d610e7730f) `1.54.0-nightly (2021-05-05 bacf770f2983a52f31e3)`

### Likely cause

It seems this behaviour is caused by [`allow_trailing_empty: false`](https://github.com/rust-lang/rust/blob/d620ae10709ca3669cda89db8b29afcd9accc188/library/core/src/str/mod.rs#L1350): `SplitInternal` only sets the finished flag if [`allow_trailing_empty: true` or there is something to return](https://github.com/rust-lang/rust/blob/d620ae10709ca3669cda89db8b29afcd9accc188/library/core/src/str/iter.rs#L581-L582).

Should the flag be set unconditionally? Maybe like this?

```rust
if !self.finished {
    self.finished = true;

    if self.allow_trailing_empty || self.end - self.start > 0 {
        return unsafe { ... } 
    }
}

None
```

### PR DESCRIPTION
**Pull request by WaffleLapkin**
I've made this patch almost a year ago, so the rename and the behavior change are in one commit, sorry 😅 

This fixes #84974, as it's required to make other changes work.

This PR
- Renames `as_str` method of string `Split*` iterators to `remainder` (it seems like the `as_str` name was confusing to users)
- Makes `remainder` return `Option<&str>`, to distinguish between "the iterator is exhausted" and "the tail is empty", this was [required on the tracking issue](https://github.com/rust-lang/rust/issues/77998#issuecomment-832696619)

r? @m-ou-se

### CODE REVIEW

File: library/core/src/str/iter.rs [hunk 1]
```diff
@@ -586,16 +586,16 @@ where
...
     fn get_end(&mut self) -> Option<&'a str> {
-        if !self.finished && (self.allow_trailing_empty || self.end - self.start > 0) {
+        if !self.finished {
             self.finished = true;
-            // SAFETY: `self.start` and `self.end` always lie on unicode boundaries.
-            unsafe {
-                let string = self.matcher.haystack().get_unchecked(self.start..self.end);
-                Some(string)
+
+            if self.allow_trailing_empty || self.end - self.start > 0 {
```
Comment (klensy) [inline library/core/src/str/iter.rs:592]: `SAFETY` comment lost.

**Comment by JohnCSimon** [comment]
@klensy @m-ou-se this is still waiting on review

**Comment by WaffleLapkin** [comment]
r? libs-api

**Pull request merged**